pennylane/fourier/spectrum.py

```diff
@@ -0,0 +1,244 @@
+    and :math:`G` is a Hermitian operator called a "generator".
+
+    If the expansion involves any processing on the parameters (such as scalar multiplication)
+    the circuit is not of a form that can be handled by the spectrum function, and an error is thrown.
+
+    Args:
+        tape (~.tape.QuantumTape): tape to simplify
+        original_inputs (list): list of inputs, to check that they are conserved throughout expansion
+
+    Returns:
+        ~.tape.QuantumTape: expanded tape
+    """
+
+    def stop_at(obj):
+        r"""Accepts a queue object and returns ``False`` if this object should be expanded.
+
+        An object should be expanded if it is not an operation, and if it is not a single-parameter gate but
+        takes inputs.
+
+        Before returning ``False`` it is checked that an expansion is possible, and that it will not perform processing
+        on the inputs.
+        """
+
+        if isinstance(obj, qml.operation.Operation):
+            takes_input = any(p in original_inputs for p in obj.parameters)
+            is_one_param_gate = obj.generator[0] is not None and len(obj.parameters) == 1
+            if takes_input and not is_one_param_gate:
+
+                # check if the trainable parameters after expansion are
+                # all found in the original inputs
+                # parameters - this makes sure no new parameters
+                # were created that depend on the inputs
+                try:
+                    expanded = obj.expand()
+                except NotImplementedError:
+                    raise ValueError(f"Cannot expand {obj}. Aborting the expansion of the tape.")
+
+                new_inputs = expanded.get_parameters(trainable_only=True)
+                if any(i not in original_inputs for i in new_inputs):
+                    raise ValueError(
+                        f"{obj} transforms the inputs. " f"Aborting the expansion of the tape."
+                    )
+
+                return False
+
+            else:
+                return True
+
+        # if object is not an op, it is probably a tape, so expand it
+        return False
+
+    return tape.expand(depth=5, stop_at=stop_at)
+
+
+def _get_spectrum(op):
+    r"""Get the spectrum of the input x of a single-parameter operation :math:`\exp(-i x G)`.
+
+    The spectrum is the set of sums of :math:`G`'s eigenvalues.
+    """
+
+    g, coeff = op.generator
+
+    # some generators are operations
+    if not isinstance(g, np.ndarray) and g is not None:
+        g = g.matrix
+
+    # the matrix or generator could be undefined
+    if g is None:
+        raise ValueError(f"no generator defined for operator {op}")
+
+    g = coeff * g
+    evals = np.linalg.eigvals(g)
+    # eigenvalues of hermitian ops are guaranteed to be real
+    evals = np.real(evals)
+
+    # append negative to cater for the complex conjugate part which subtracts eigenvalues
+    evals = [evals, -evals]
```

Comment:
I have to admit I am a bit confused why this is right here...I stole this from you

Reply:
I believe this is from your original version of the function :sweat_smile:

Reply:
Oh shoot :) No, I remember. Each gate appears in the left and right side of the quadratic form that makes up the expval!